abci/proposals/proposals.go

```diff
@@ -85,45 +136,70 @@ func (h *ProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
 			)
 
 			extInfo := req.LocalLastCommit
-			if err := h.ValidateExtendedCommitInfo(ctx, req.Height, extInfo); err != nil {
+			if err = h.ValidateExtendedCommitInfo(ctx, req.Height, extInfo); err != nil {
 				h.logger.Error(
 					"failed to validate vote extensions",
 					"height", req.Height,
 					"commit_info", extInfo,
 					"err", err,
 				)
+				err = InvalidExtendedCommitInfoError{
+					Err: err,
+				}
 
 				return &cometabci.ResponsePrepareProposal{Txs: make([][]byte, 0)}, err
 			}
 
-			// Inject the vote extensions into the proposal. These contain the oracle data
-			// for the current block which will be committed to state in PreBlock.
-			extInfoBz, err = extInfo.Marshal()
+			// Create the vote extension injection data which will be injected into the proposal. These contain the
+			// oracle data for the current block which will be committed to state in PreBlock.
+			extInfoBz, err = h.extendedCommitCodec.Encode(extInfo)
 			if err != nil {
 				h.logger.Error(
 					"failed to extended commit info",
 					"commit_info", extInfo,
 					"err", err,
 				)
+				err = slinkyabci.CodecError{
+					Err: err,
+				}
 
 				return &cometabci.ResponsePrepareProposal{Txs: make([][]byte, 0)}, err
 			}
+			// Adjust req.MaxTxBytes to account for extInfoBzSize so that the wrapped-proposal handler does not reap too many txs from the mempool
+			extInfoBzSize := int64(len(extInfoBz))
+			if extInfoBzSize < req.MaxTxBytes {
+				// Reserve bytes for our VE Tx
+				req.MaxTxBytes -= extInfoBzSize
+			} else {
+				h.logger.Error("omitting VE because size consumes entire block",
+					"extInfoBzSize", extInfoBzSize,
+					"MaxTxBytes", req.MaxTxBytes)
+				extInfoBz = []byte{}
+			}
 
-			req.Txs = append([][]byte{extInfoBz}, req.Txs...)
+			// determine whether the wrapped prepare proposal handler should retain the extended commit info
+			if h.retainOracleDataInWrappedHandler {
+				req.Txs = append([][]byte{extInfoBz}, req.Txs...) // prepend the VE Tx
+			}
 		}
 
-		// Build the proposal.
-		resp, err := h.prepareProposalHandler(ctx, req)
+		// Build the proposal. Get the duration that the wrapped prepare proposal handler executed for.
+		wrappedPrepareProposalStartTime := time.Now()
+		resp, err = h.prepareProposalHandler(ctx, req)
+		wrappedPrepareProposalLatency = time.Since(wrappedPrepareProposalStartTime)
 		if err != nil {
 			h.logger.Error("failed to prepare proposal", "err", err)
+			err = slinkyabci.WrappedHandlerError{
+				Handler: servicemetrics.PrepareProposal,
+				Err:     err,
+			}
+
 			return &cometabci.ResponsePrepareProposal{Txs: make([][]byte, 0)}, err
```

Comment:
https://informalsystems.atlassian.net/wiki/spaces/SS/pages/164986910/Incorrect+defensive+coding+in+ABCI+functions+if+req+is+nil